tests/test_urldispatch.py

```diff
@@ -770,13 +770,24 @@ async def test_dynamic_match_two_part2(router: web.UrlDispatcher) -> None:
 
 async def test_dynamic_match_unquoted_path(router: web.UrlDispatcher) -> None:
     handler = make_handler()
-    router.add_route("GET", "/{path}/{subpath}", handler)
```

Comment:
I think if this change is required, then the issue I fixed is now not fixed...

I don't think this is properly fixable without have separate attributes in yarl or something...

To distill the problem down to the minimal issue:
Problem with old yarl:
`URL("/", encoded=True).path == URL("%2F", encoded=True).path`
Problem with new yarl:
`URL("%2F", encoded=True).path == URL("%252F", encoded=True).path`

I don't see any way to fix that while having only one path attribute to handle it.

Reply:
I'm not sure that will take care of this case

The regex now runs with unquoted path at https://github.com/aio-libs/aiohttp/blob/203ca4273edcd6ea5372c630b054381ebb3bacdb/aiohttp/web_urldispatcher.py#L473

With the yarl change, %2F is never unquoted

So if when return the `match.groupdict()` its going to have `%2F` if there was a quoted `/`

We could replace `%2F` with `/` but that means that we would also convert `%252F` to `/`

Reply:
> We don't have any way to convert the match back to the raw url parts and than unquote them after we get the match result

Exactly, the problem exists with both new and old yarl though, it's just different situation. The only way I can think of, is to have 2 different attributes which are used here. e.g. Using the new .path for routing and the old .path for matchinfo.

Reply:
I guess the matchinfo is part of the routing though, so that's not so easy...

Reply:
Yeah we have to return the match dict and once we decode it we can never know if the / was encoded in the original as `%2F` or `%252F`

Reply:
```
> URL('%2F', encoded=True).path
'%2F'
> URL('%252F', encoded=True).path
'%2F'
```

Reply:
Yeah, that's what I highlighted in the first comment. Likewise, we can't tell if the `/` was a path separator or not with the old yarl for the same reason.

Reply:
In theory, I think we could do to matches, although that's might not be best for performance:
```
    def _match(self, path: str, unslashed_path: str) -> Optional[Dict[str, str]]:
        match = self._pattern.fullmatch(path)  # new-style path for routing
        if match is None:
            return None

        match = self._pattern.fullmatch(unslashed_path)  # old-style path for matchinfo
        return {key: _unquote_path(value) for key, value in match.groupdict().items()}
```

Reply:
I suppose the pattern would need to be changed too though...

Reply:
What about not unquoting `%25` and then adding that to the .replace()? I think that could work, though it might make things more awkward for other users of yarl..

Reply:
So, maybe we end up with `url.path_safe.replace("%2F", "/").replace("%25", "%")` or similar.

Reply:
It subtly affect the regex matching, as it'd now be matching against `%25` instead of `%`, but I'm not sure of any better way.

Reply:
pushed a change that seems to work ok with https://github.com/aio-libs/yarl/pull/1150